Add autosave feature

superenginegit opened the pull request:
This PR is for the feature request of issue #1926.

**What I implemented:**
- Settings for autosaves: Autosaving can be toggled on or off and the time interval between two autosaves can be set.
- Creation of autosaves: Autosaving is performed regularly, but a new autosave will only be created if there were changes since the last autosave. Autosaves will be made for files that have been saved, for those they will be created in the same location as the original file. Autosave for files that have never be saved will be created in the users home directory.
- Autosaves that have been created for unsaved files will be checked for on start of logisim. And they can then be saved.
- Autosaves for saved files will be checked when that file is reopened. It can then be restored or deleted.

maehne commented on src/main/java/com/cburch/logisim/file/Loader.java:
```diff
@@ -142,6 +149,28 @@ private static File determineBackupName(File base) {
+  private static File determineAutosaveName(File base) {
+    if (base == null) {
+      String timestamp =
+          DateTimeFormatter.ofPattern("yyyy-MM-dd_HH:mm:ss").format(LocalDateTime.now());
```
`:` in file names are not universally allowed, e.g., Windows doesn't allow them. Personally, I would opt to use `yyyyMMddHHmmss`, which also enables chronological sorting.

maehne commented on src/main/java/com/cburch/logisim/file/Loader.java:
```diff
@@ -419,9 +449,50 @@ public boolean save(LogisimFile file, File dest) {
+   *
+   * @param file The file that should be autosaved
+   *
+   * @return True if writing was successfull, else false;
```
Typo: "successfull" -> "successful"

maehne commented on src/main/java/com/cburch/logisim/file/LogisimFile.java:
```diff
@@ -163,10 +214,32 @@ private static String getFirstLine(BufferedInputStream in) throws IOException {
+      } else if (res == 0) { // If load is selected select the autosave to be loaded
+        loadFile = autosave.get(); // Set load file to the autosave path
+        loader.setAutosavePath(autosave.get());
+        autosaveLoading = true; // also set this to true to remember an autosave was loaded
```
"also" -> "Also" to be consistent with casing in other comments?

maehne commented on src/main/resources/resources/logisim/strings/file/file.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Library includes items currently in the toolbar.
 unloadUsedError = Circuit '%s' uses components from the library.
 xmlConversionError = Internal error while creating XML
 xmlFormatError = XML formatting error: %s
+autosaveError = Failed to create autosave for file: '%s'. Will not try again for this File until after a restart.
```
"File" -> "file"

maehne reviewed (changes requested):
Thanks for implementing this useful feature! It looks mostly good for me, but as this modifies some quite sensitive parts of the code, it should be at least also reviewed by @BFH-ktt1. It would be nice if you could add the newly introduced localisation strings to alls supported languages as placeholders at least. This facilitates additional translations.

superenginegit commented:
Thanks for the review, I have now added translation placeholders to the languages and fixed the other things you mentioned as well.

maehne commented on src/main/resources/resources/logisim/strings/file/file_fr.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
 unloadUsedError = Le circuit '%s' utilise des composants de la librarie.
 xmlConversionError = Erreur interne pendant la création du fichier XML
 xmlFormatError = Erreur de formatage XML : %s
+# ==> autosaveError = Failed to create autosave for file: '%s'. Will not try again for this file until after a restart.
```
"Échec de la création d'une sauvegarde automatique pour le fichier : '%s'. La tentative ne sera pas renouvelée pour ce fichier avant un redémarrage."

maehne commented on src/main/resources/resources/logisim/strings/file/file_fr.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
 xmlConversionError = Erreur interne pendant la création du fichier XML
 xmlFormatError = Erreur de formatage XML : %s
+# ==> autosaveError = Failed to create autosave for file: '%s'. Will not try again for this file until after a restart.
+# ==> titleHandleAutosave = Autosave found
```
"Sauvegarde automatique trouvée"

maehne commented on src/main/resources/resources/logisim/strings/file/file_fr.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
 xmlFormatError = Erreur de formatage XML : %s
+# ==> autosaveError = Failed to create autosave for file: '%s'. Will not try again for this file until after a restart.
+# ==> titleHandleAutosave = Autosave found
+# ==> contentHandleAutosave = Autosave for file '%s' found. What should be done?
```
"Sauvegarde automatique pour le fichier '%s' trouvée. Que faut-il faire ?"

maehne commented on src/main/resources/resources/logisim/strings/file/file_fr.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
+# ==> autosaveError = Failed to create autosave for file: '%s'. Will not try again for this file until after a restart.
+# ==> titleHandleAutosave = Autosave found
+# ==> contentHandleAutosave = Autosave for file '%s' found. What should be done?
+# ==> loadOption = Load
```
"Charger"

maehne commented on src/main/resources/resources/logisim/strings/file/file_fr.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = La librairie inclut des éléments de la barre d'outils.
+# ==> titleHandleAutosave = Autosave found
+# ==> contentHandleAutosave = Autosave for file '%s' found. What should be done?
+# ==> loadOption = Load
+# ==> discardOption = discard
```
"Écarter"

maehne commented on src/main/resources/resources/logisim/strings/file/file_nl.properties:
@BFH-ktt1: Maybe you could provide translations for these strings?

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
 #
+# prefs/AutosaveOptions.java
+#
+# ==> autosaveEnabled = Enable autosaving (Takes effect after a restart).
```
"Activer la sauvegarde automatique (prend effet après un redémarrage)."

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
+# prefs/AutosaveOptions.java
+#
+# ==> autosaveEnabled = Enable autosaving (Takes effect after a restart).
+# ==> autosaveInterval = Autosaving interval in seconds:
```
"Intervalle de sauvegarde automatique en secondes :"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
+#
+# ==> autosaveEnabled = Enable autosaving (Takes effect after a restart).
+# ==> autosaveInterval = Autosaving interval in seconds:
+# ==> autosaveTitle = Autosave
```
"Sauvegarde automatique"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Sélectionner...
+# ==> autosaveEnabled = Enable autosaving (Takes effect after a restart).
+# ==> autosaveInterval = Autosaving interval in seconds:
+# ==> autosaveTitle = Autosave
+# ==> autosaveHelp = Select autosave settings
```
"Sélectionner les paramètres de sauvegarde automatique"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
 templateMissingError = Le fichier modèle %s n'existe pas.
 ttyFormatError = --tty requière au moins l'une des options : halt, speed, stats, table, tty, binary, hex, csv, tabs
 ttyNeedsFileError = L'utilisation de "--tty" requière un nom de fichier sur la ligne de commande.
+# ==> saveAndLoadOption = Save and load
```
"Sauvegarder et charger"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
 ttyFormatError = --tty requière au moins l'une des options : halt, speed, stats, table, tty, binary, hex, csv, tabs
 ttyNeedsFileError = L'utilisation de "--tty" requière un nom de fichier sur la ligne de commande.
+# ==> saveAndLoadOption = Save and load
+# ==> titleHandleAutosave = Autosave found
```
"Sauvegarde automatique trouvée"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_fr.properties:
```diff
@@ -743,6 +750,9 @@ templateCannotReadError = Pas la permission de lire le fichier modèle %s.
 ttyNeedsFileError = L'utilisation de "--tty" requière un nom de fichier sur la ligne de commande.
+# ==> saveAndLoadOption = Save and load
+# ==> titleHandleAutosave = Autosave found
+# ==> contentHandleAutosave = Autosave file '%s' found. What should be done?
```
"Fichier de sauvegarde automatique '%s' trouvé. Que faut-il faire ?"

maehne commented on src/main/resources/resources/logisim/strings/file/file_de.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
 unloadUsedError = Schaltung "%s" verwendet Elemente der Bibliothek.
 xmlConversionError = Interner Fehler beim Erstellen von XML
 xmlFormatError = Fehler beim XML-Format: %s
+autosaveError = Erstellen der Autospeicherung für Datei '%s' fehlgeschlagen. Wird bis zum Neustart nicht wiederholt.
```
"Automatische Sicherung für Datei konnte nicht erstellt werden: '%s'. Wird erst nach einem Neustart erneut für diese Datei versucht."

maehne commented on src/main/resources/resources/logisim/strings/file/file_de.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
 xmlConversionError = Interner Fehler beim Erstellen von XML
 xmlFormatError = Fehler beim XML-Format: %s
+autosaveError = Erstellen der Autospeicherung für Datei '%s' fehlgeschlagen. Wird bis zum Neustart nicht wiederholt.
+titleHandleAutosave = Autospeicherung gefunden
```
"Automatische Sicherung gefunden"

maehne commented on src/main/resources/resources/logisim/strings/file/file_de.properties:
```diff
@@ -44,6 +44,11 @@ unloadToolbarError = Bibliothek enthält Elemente auf der Werkzeugleiste.
 xmlFormatError = Fehler beim XML-Format: %s
+autosaveError = Erstellen der Autospeicherung für Datei '%s' fehlgeschlagen. Wird bis zum Neustart nicht wiederholt.
+titleHandleAutosave = Autospeicherung gefunden
+contentHandleAutosave = Autospeicherung für Datei '%s' gefunden. Wie vorgehen?
```
Automatische Sicherung für Datei '%s' gefunden. Wie vorgehen?

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_de.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
 #
+# prefs/AutosaveOptions.java
+#
+autosaveEnabled = Aktiviere Autospeicherungen (Erfordert Neustart).
```
Aktiviere automatische Sicherungen (erfordert Neustart).

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_de.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
+#
+autosaveEnabled = Aktiviere Autospeicherungen (Erfordert Neustart).
+autosaveInterval = Autospeicherungsinterval in Sekunden:
+autosaveTitle = Autospeichern
```
"Automatische Sicherungen"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_de.properties:
```diff
@@ -600,6 +600,13 @@ templateSelectButton = Auswählen...
+autosaveEnabled = Aktiviere Autospeicherungen (Erfordert Neustart).
+autosaveInterval = Autospeicherungsinterval in Sekunden:
+autosaveTitle = Autospeichern
+autosaveHelp = Autospeicherungseinstellungen ündern
```
"Einstellungen für die automatischen Sicherungen ändern"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_de.properties:
```diff
@@ -743,6 +750,9 @@ startupQuitButton = Beenden
 # ==> ttyFormatError =
 # ==> ttyNeedsFileError =
+saveAndLoadOption = Speichern und laden
+titleHandleAutosave = Autospeicherung gefunden
```
"Automatische Sicherung gefunden"

maehne commented on src/main/resources/resources/logisim/strings/gui/gui_de.properties:
```diff
@@ -743,6 +750,9 @@ startupQuitButton = Beenden
 # ==> ttyNeedsFileError =
+saveAndLoadOption = Speichern und laden
+titleHandleAutosave = Autospeicherung gefunden
+contentHandleAutosave = Autospeicherungsdatei '%s' gefunden. Wie vorgehen?
```
Automatische Sicherungsdatei '%s' gefunden. Wie vorgehen?

maehne reviewed (changes requested):
Thanks for taking into account my review comments! I suggest some further improvements to the translations.

davidhutchens reviewed (commented):
I looked through the code and ran some tests to see how this worked. It looks good to me. Thank you for this contribution. I have one small request.

Would you please add a line to CHANGES.md to note the addition of an autosave feature?

superenginegit commented:
I've implemented the translation strings you've provided. And I've also added a line to the changes file noting the addition of autosaves. Thanks for your feedback!

maehne reviewed (approved):
From my side this PR is ready for merging. Unless @BFH-ktt1 would still like to provide Dutch localization strings.
